actiontext/lib/action_text/attachable.rb

```diff
@@ -99,8 +99,9 @@ def previewable_attachable?
     end
 
     # Returns the attachable as JSON with the +attachable_sgid+ included.
-    def as_json(*)
-      super.merge("attachable_sgid" => persisted? ? attachable_sgid : nil)
+    def as_json(*args)
+      options = args[0].to_h.transform_values { |v| Array(v).map(&:to_s) if v }
+      super.merge("attachable_sgid" => persisted? ? attachable_sgid : nil).as_json(options)
     end
```

Comment:
Instead of changing the `as_json` implementation, I wonder if we'd be better off fixing this behavior with a different approach.

Calling `#as_json` is the final step in a chain of method calls during the JSON serialization process. Prior to `as_json`, objects will call [#serializable_hash](https://edgeapi.rubyonrails.org/classes/ActiveModel/Serialization.html#method-i-serializable_hash) with the same `options = nil` argument that `#as_json` accepts.

That method calls two other methods to determine the keys and values to return:

1. [#attribute_names_for_serialization](https://github.com/rails/rails/blob/7c9eac6923ab5f39c04d03f6c3a3ae3c99ce42d5/activemodel/lib/active_model/serialization.rb#L152-L154)
2. [#read_attribute_for_serialization](https://github.com/rails/rails/blob/7c9eac6923ab5f39c04d03f6c3a3ae3c99ce42d5/activemodel/lib/active_model/serialization.rb#L156-L172)

If we override those methods to incorporate the `attachable_sgid` method, we can avoid managing the `#as_json` method entirely:

```diff
diff --git a/actiontext/lib/action_text/attachable.rb b/actiontext/lib/action_text/attachable.rb
index 4e2e20db14..b15b10418c 100644
--- a/actiontext/lib/action_text/attachable.rb
+++ b/actiontext/lib/action_text/attachable.rb
@@ -98,12 +98,6 @@ def previewable_attachable?
       false
     end
 
-    # Returns the attachable as JSON with the +attachable_sgid+ included.
-    def as_json(*args)
-      options = args[0].to_h.transform_values { |v| Array(v).map(&:to_s) if v }
-      super.merge("attachable_sgid" => persisted? ? attachable_sgid : nil).as_json(options)
-    end
-
     # Returns the path to the partial that is used for rendering the attachable
     # in Trix. Defaults to +to_partial_path+.
     #
@@ -143,5 +137,19 @@ def to_rich_text_attributes(attributes = {})
         attrs[:height] = attachable_metadata[:height]
       end.compact
     end
+
+    private
+
+      def attribute_names_for_serialization
+        super + ["attachable_sgid"]
+      end
+
+      def read_attribute_for_serialization(key)
+        if key.to_s == "attachable_sgid"
+          persisted? ? super : nil
+        else
+          super
+        end
+      end
   end
 end

```

Reply:
@seanpdoyle 
Thanks again! 
Your way appears more refined, and I've made the changes.
I'd appreciate it if you could review it again when you have the time.